### PR TITLE
chore: update CredentialProvider

### LINKED ISSUE #100: Update CredentialProvider to include string provider and helper methods
The PHP SDK currently only implements the `EnvMomentoTokenProvider` Credential Provider. Other SDKs have introduced a `StringMomentoTokenProvider` that we should port over to PHP. The `fromString` and `fromEnvironmentVariable` static methods that have been introduced in the other SDKs also need to be ported over.

### PR DESCRIPTION
**Pull request by poppoerika**
- Added `StringMomentoTokenProvider`
- Added static functions, `fromEnvironmentVariable` and `fromString` 
- Added some test to test out the above

Closes #100

### CODE REVIEW

**Comment by malandis** [comment]
Looks good. Can you document the `StringMomentoProvider` class and `fromEnvironmentVariable` method as well?

**Comment by pgautier404** [comment]
This looks good but there is a lot of duplicated code between the two providers. In other SDKs we had the `EnvMomentoTokenProvider` inherit from `StringMomentoTokenProvider` since the only extra thing it does is read a stering from an environment variable. I think doing the same here would simplify and make this more maintainable. The node SDK is a good model for this pattern.

**Comment by poppoerika** [comment]
> This looks good but there is a lot of duplicated code between the two providers. In other SDKs we had the `EnvMomentoTokenProvider` inherit from `StringMomentoTokenProvider` since the only extra thing it does is read a stering from an environment variable. I think doing the same here would simplify and make this more maintainable. The node SDK is a good model for this pattern.

@pgautier404 
I wanted to discuss this with you, so sounds good!

File: src/Auth/ICredentialProvider.php [hunk 1]
```diff
@@ -1,17 +0,0 @@
-<?php
-declare(strict_types=1);
-
-namespace Momento\Auth;
-
-interface ICredentialProvider
```
Comment (cprice404) [inline src/Auth/ICredentialProvider.php:6]: why delete the interface?

**Review by cprice404 — COMMENTED** [review]
i feel like it's probably still a good idea to keep the interface around and use it in the return types from the factory functions, but will defer to Pete.

File: src/Auth/ICredentialProvider.php (hunk 1 above)
Comment (pgautier404) [reply src/Auth/ICredentialProvider.php:6]: Yeah, we kind of went back and forth on this but I think @cprice404 is right. Keeping the interface around gives us extra flexibility, and that's definitely what we should be returning from the factory methods, @poppoerika.

File: src/Auth/CredentialProvider.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?php
+declare(strict_types=1);
+
+namespace Momento\Auth;
+
+/**
+ * Provides information that the CacheClient needs in order to establish a connection to and authenticate with
+ * the Momento service.
+ */
+abstract class CredentialProvider
```
Comment (pgautier404) [inline src/Auth/CredentialProvider.php:10]:
```suggestion
abstract class CredentialProvider implements ICredentialProvider
```

We'll of course need to add the `ICredentialProvider` file and class back in.

File: src/Auth/EnvMomentoTokenProvider.php [hunk 3]
```diff
@@ -27,42 +23,17 @@ public function __construct(
...
     }
 
-    public function getTrustedCacheEndpointCertificateName(): string|null
+    /**
+     * Convenience method for reading and parsing a JWT token stored as an environment variable.
+     * @param string $envVariableName Name of the environment variable that contains the JWT token.
+     * @return EnvMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromEnvironmentVariable(string $envVariableName): EnvMomentoTokenProvider
```
Comment (pgautier404) [inline src/Auth/EnvMomentoTokenProvider.php:35]: This function should be in the `CredentialProvider` class and should return `ICredentialProvider` type.

File: src/Auth/StringMomentoTokenProvider.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        return $this->trustedCacheEndpointCertificateName;
+    }
+
+    /**
+     * Convenience method for reading and parsing a JWT token stored as a string.
+     * @param string $authToken The JWT token.
+     * @return StringMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromString(string $authToken): StringMomentoTokenProvider
```
Comment (pgautier404) [inline src/Auth/StringMomentoTokenProvider.php:70]: Please move this function to `CredentialProvider` and mark it as returning an `ICredentialProvider`.

File: src/Cache/CacheClient.php [hunk 5]
```diff
@@ -48,11 +48,11 @@ class CacheClient implements LoggerAwareInterface
 
     /**
      * @param IConfiguration $configuration
-     * @param ICredentialProvider $authProvider
+     * @param CredentialProvider $authProvider
      * @param int $defaultTtlSeconds
      */
     public function __construct(
-        IConfiguration $configuration, ICredentialProvider $authProvider, int $defaultTtlSeconds
+        IConfiguration $configuration, CredentialProvider $authProvider, int $defaultTtlSeconds
```
Comment (pgautier404) [inline src/Cache/CacheClient.php:55]:
```suggestion
        IConfiguration $configuration, ICredentialProvider $authProvider, int $defaultTtlSeconds
```

Comment (pgautier404) [reply src/Cache/CacheClient.php:55]: Sorry about the runaround, but we'll want these types to be `ICredentialProvider` pretty much everywhere in the clients and grpc managers.

**Comment by poppoerika** [comment]
@pgautier404 @cprice404 
Thank you for the feedback! 
After talking to Pete, it makes sense to keep `ICredentialProvider` and implement it in `CredentialProvider` to add flexibility to our interface.
I added back `ICredentialProvider` and updated `CredentialProvier` accordingly. 
Also updated the examples so that we use `CacheClient` and `CredentialProvider::fromEnvironmentVariable`.
Ah, also updated `CredentialProvider` return type to `ICredentialProvier` type.

File: src/Cache/Psr16CacheClient.php [hunk 6]
```diff
@@ -26,15 +26,15 @@ class Psr16CacheClient implements CacheInterface
 
     /**
      * @param IConfiguration $configuration
-     * @param ICredentialProvider $authProvider
+     * @param CredentialProvider $authProvider
```
Comment (pgautier404) [inline src/Cache/Psr16CacheClient.php:29]:
```suggestion
     * @param ICredentialProvider $authProvider
```

File: src/Cache/Psr16CacheClient.php [hunk 7]
```diff
@@ -26,15 +26,15 @@ class Psr16CacheClient implements CacheInterface
...
      * @param int|null $defaultTtlSeconds
      * @param bool|null $throwExceptions
      */
     public function __construct(
-        IConfiguration      $configuration,
-        ICredentialProvider $authProvider,
-        ?int                $defaultTtlSeconds,
-        ?bool               $throwExceptions = null
+        IConfiguration     $configuration,
+        CredentialProvider $authProvider,
```
Comment (pgautier404) [inline src/Cache/Psr16CacheClient.php:35]:
```suggestion
        ICredentialProvider $authProvider,
```

File: src/Auth/CredentialProvider.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    public abstract function getTrustedCacheEndpointCertificateName(): string|null;
+
+    /**
+     * Convenience method for reading and parsing a JWT token stored as a string.
+     * @param string $authToken The JWT token.
+     * @return StringMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromString(string $authToken): StringMomentoTokenProvider
```
Comment (pgautier404) [inline src/Auth/CredentialProvider.php:26]:
```suggestion
    public static function fromString(string $authToken): ICredentialProvider
```

File: src/Auth/CredentialProvider.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        return new StringMomentoTokenProvider($authToken);
+    }
+
+    /**
+     * Convenience method for reading and parsing a JWT token stored as an environment variable.
+     * @param string $envVariableName Name of the environment variable that contains the JWT token.
+     * @return EnvMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromEnvironmentVariable(string $envVariableName): EnvMomentoTokenProvider
```
Comment (pgautier404) [inline src/Auth/CredentialProvider.php:37]:
```suggestion
    public static function fromEnvironmentVariable(string $envVariableName): ICredentialProvider
```

**Review by pgautier404 — APPROVED** [review]
LGTM!

**Pull request merged**